apps/remote_control/lib/lexical/remote_control/search/indexer/extractors/module.ex

```diff
@@ -4,137 +4,111 @@ defmodule Lexical.RemoteControl.Search.Indexer.Extractors.Module do
 
-    module_name_meta = Reducer.skip(module_name_meta)
+  @impl Extractor
+  def extract(zipper, extractor)
 
-    elem =
-      {:defmodule, defmodule_meta, [{:__aliases__, module_name_meta, module_name}, module_block]}
-
-    {:ok, entry, elem}
-  end
-
-  # This matches an elixir module reference
-  def extract({:__aliases__, metadata, maybe_module}, %Reducer{} = reducer)
-      when is_list(maybe_module) do
-    case module(reducer, maybe_module) do
+  # Elixir module definition or reference
+  def extract(
+        %Zipper{node: {:__aliases__, _, segments}} = zipper,
+        %Extractor{} = extractor
+      ) do
+    case fetch_module(segments, extractor) do
       {:ok, module} ->
-        start = Metadata.position(metadata)
-        range = to_range(reducer, maybe_module, start)
-        %Block{} = current_block = Reducer.current_block(reducer)
-
-        entry =
-          Entry.reference(
-            reducer.analysis.document.path,
-            make_ref(),
-            current_block.ref,
-            module,
-            :module,
-            range,
-            Application.get_application(module)
-          )
-
-        {:ok, entry}
-
-      _ ->
-        :ignored
+        extract_module(extractor, module, zipper)
+
+      :error ->
+        extractor
     end
   end
 
-  # This matches an erlang module, which is just an atom
-  def extract({:__block__, metadata, [atom_literal]}, %Reducer{} = reducer)
+  # Erlang module definition or reference, which are plain atoms
+  def extract(
+        %Zipper{node: {:__block__, _, [atom_literal]}} = zipper,
+        %Extractor{} = extractor
+      )
       when is_atom(atom_literal) do
-    case module(reducer, atom_literal) do
+    case fetch_erlang_module(atom_literal) do
       {:ok, module} ->
-        start = Metadata.position(metadata)
-        %Block{} = current_block = Reducer.current_block(reducer)
-        range = to_range(reducer, module, start)
-
-        entry =
-          Entry.reference(
-            reducer.analysis.document.path,
-            make_ref(),
-            current_block.ref,
-            module,
-            :module,
-            range,
-            Application.get_application(module)
-          )
-
-        {:ok, entry}
+        extract_module(extractor, module, zipper)
 
       :error ->
-        :ignored
+        extractor
     end
   end
 
-  def extract(_, _) do
-    :ignored
+  def extract(_zipper, %Extractor{} = extractor) do
+    extractor
   end
 
-  defp resolve_alias(%Reducer{} = reducer, unresolved_alias) do
-    {line, column} = reducer.position
-    position = Position.new(reducer.analysis.document, line, column)
-
-    {:ok, expanded} = Ast.expand_alias(unresolved_alias, reducer.analysis, position)
-
-    expanded
+  defp extract_module(%Extractor{} = extractor, module, %Zipper{} = zipper) do
+    {zipper, subtype, parent_kind} =
+      case Zipper.up(zipper) do
+        %Zipper{node: {:defmodule, _, [_, _]}} = defmodule_zipper ->
```

Comment:
Using a zipper means I can easily determine whether the current position, when traversing the simplest target AST, is a definition or a reference (because it allows easy access to parent or sibling nodes, right?

If that's the case, it's simply amazing.

Reply:
Zippers are cool, yes, but they incur a fairly steep performance penalty over the native methods for AST traversal.

Reply:
I was thinking about this as well -- we may be able to get away with just having access to ancestors. This still has a lot of benefits and would work, I think, in place of all the zipper usage in this PR, and also would be helpful for @scottming's PR.